src/libos/src/net/syscalls.rs

```diff
@@ -135,6 +135,126 @@ pub async fn do_accept4(
     Ok(new_fd as isize)
 }
 
+pub async fn do_sendto(
+    fd: c_int,
+    base: *const c_void,
+    len: size_t,
+    flags: c_int,
+    addr: *const libc::sockaddr,
+    addr_len: libc::socklen_t,
+) -> Result<isize> {
+    if len == 0 {
+        return Ok(0);
+    }
+
+    let file_ref = current!().file(fd as FileDesc)?;
+    let socket_file = file_ref
+        .as_socket_file()
+        .ok_or_else(|| errno!(EINVAL, "not a socket"))?;
+
+    let flags = SendFlags::from_bits_truncate(flags);
+
+    from_user::check_array(base as *const u8, len)?;
+    let buf = unsafe { std::slice::from_raw_parts(base as *const u8, len as usize) };
+
+    let addr = if socket_file.is_stream() {
+        if !addr.is_null() || addr_len != 0 {
+            return_errno!(EISCONN, "addr and addr_len should be null and 0");
+        }
+        None
+    } else {
+        if addr.is_null() || addr_len == 0 {
+            return_errno!(EINVAL, "addr or addr_len is invalid");
+        }
+        let addr_storage = copy_sock_addr_from_user(addr, addr_len as _)?;
+        Some(AnyAddr::from_c_storage(&addr_storage, addr_len as usize)?)
+    };
+
+    socket_file
+        .sendto(&buf, addr, flags)
+        .await
+        .map(|bytes_send| bytes_send as isize)
+}
+
+pub async fn do_recvfrom(
+    fd: c_int,
+    base: *mut c_void,
+    len: size_t,
+    flags: c_int,
+    addr: *mut libc::sockaddr,
+    addr_len: *mut libc::socklen_t,
+) -> Result<isize> {
+    if addr.is_null() ^ addr_len.is_null() {
+        return_errno!(EINVAL, "addr and addr_len should be both null or not null");
+    }
+
+    let file_ref = current!().file(fd as FileDesc)?;
+    let socket_file = file_ref
+        .as_socket_file()
+        .ok_or_else(|| errno!(EINVAL, "not a socket"))?;
+
+    let flags = RecvFlags::from_bits_truncate(flags);
+
+    from_user::check_array(base as *mut u8, len)?;
+    let mut buf = unsafe { std::slice::from_raw_parts_mut(base as *mut u8, len as usize) };
+
+    let (bytes_recv, addr_recv) = socket_file.recvfrom(&mut buf, flags).await?;
+    if !addr.is_null() && addr_recv.is_some() {
+        let (c_addr_storage, c_addr_len) = addr_recv.unwrap().to_c_storage();
+        copy_sock_addr_to_user(c_addr_storage, c_addr_len, addr, addr_len)?;
```

Comment:
I get the idea from linux syscall, since linux syscall will do all things and then do copy_to_user, which might failed.
I agree that extracting the pointer checking logic and putting in the beginning of syscall